src/python/pants/backend/python/rules/hermetic_pex.py

```diff
@@ -1,64 +1,117 @@
-from pants.engine.fs import Digest
 from pants.engine.isolated_process import ExecuteProcessRequest
+from pants.engine.rules import RootRule, rule, subsystem_rule
 from pants.python.python_setup import PythonSetup
+from pants.subsystem.subsystem import Subsystem
 from pants.util.strutil import create_path_env_var
 
 
-class HermeticPex:
-    """A mixin for types that provide an executable Pex that should be executed hermetically."""
-
-    def create_execute_request(
-        self,
-        python_setup: PythonSetup,
-        subprocess_encoding_environment: SubprocessEncodingEnvironment,
-        *,
-        pex_path: str,
-        pex_args: Iterable[str],
-        description: str,
-        input_files: Digest,
-        env: Optional[Mapping[str, str]] = None,
-        **kwargs: Any
-    ) -> ExecuteProcessRequest:
-        """Creates an ExecuteProcessRequest that will run a PEX hermetically.
-
-        :param python_setup: The parameters for selecting python interpreters to use when invoking
-                             the PEX.
-        :param subprocess_encoding_environment: The locale settings to use for the PEX invocation.
-        :param pex_path: The path within `input_files` of the PEX file (or directory if a loose
-                         pex).
-        :param pex_args: The arguments to pass to the PEX executable.
-        :param description: A description of the process execution to be performed.
-        :param input_files: The files that contain the pex itself and any input files it needs to
-                            run against.
-        :param env: The environment to run the PEX in.
-        :param **kwargs: Any additional :class:`ExecuteProcessRequest` kwargs to pass through.
-        """
-
-        # NB: we use the hardcoded and generic bin name `python`, rather than something dynamic like
-        # `sys.executable`, to ensure that the interpreter may be discovered both locally and in remote
-        # execution (so long as `env` is populated with a `PATH` env var and `python` is discoverable
-        # somewhere on that PATH). This is only used to run the downloaded PEX tool; it is not
-        # necessarily the interpreter that PEX will use to execute the generated .pex file.
-        # TODO(#7735): Set --python-setup-interpreter-search-paths differently for the host and target
-        # platforms, when we introduce platforms in https://github.com/pantsbuild/pants/issues/7735.
-        argv = ("python", pex_path, *pex_args)
-
-        hermetic_env = dict(
-            PATH=create_path_env_var(python_setup.interpreter_search_paths),
-            PEX_ROOT="./pex_root",
-            PEX_INHERIT_PATH="false",
-            PEX_IGNORE_RCFILES="true",
-            **subprocess_encoding_environment.invocation_environment_dict
-        )
-        if env:
-            hermetic_env.update(env)
+@dataclass(frozen=True)
+class HermeticPexRequest:
```

Comment:
Consider adding a `create()` factory method with a similar signature to `create_execute_request`, but without the fields that are now unnecessary.

I thought that it read nicely to say `requirements_pex.create_execute_request(pex_path="./black.pex", pex_args=[])` because it made it extremely explicit that you will always have those two values when dealing with PEXes, rather than having the user need to figure out they should manually construct `["./black.pex", "arg1"]`.

You could then also move the `env` handling into the factory method, which is probably closer to where it belongs per the principle of proximity.

Reply:
Why do we want to simulate the fields of `ExecuteProcessRequest` in kwargs instead of accepting a process execution request? I thought that was one of the most confusing things to follow from the previous solution and made things significantly more magical/mysterious. There **is** a real `ExecuteProcessRequest` being created here, and we're just passing kwargs to it -- why hide this?

> I thought that it read nicely to say requirements_pex.create_execute_request(pex_path="./black.pex", pex_args=[]) because it made it extremely explicit that you will always have those two values when dealing with PEXes, rather than having the user need to figure out they should manually construct ["./black.pex", "arg1"].

That also introduced the possibility of making a mistake and using the wrong `pex_path`, which is another thing the current solution avoids. The user doesn't have to manually construct anything in the current solution, so I'm not sure I understand what benefit you're trying to describe here.

> You could then also move the env handling into the factory method, which is probably closer to where it belongs per the principle of proximity.

I don't understand what the principle of proximity means here. We create the `ExecuteProcessRequest` to be executed in the new `@rule`. I don't understand the benefit of moving that logic out of the rule, when the new `@rule` is the only thing consuming the env handling?

Reply:
Ah, I see on a re-read that you are not having the user have to manually specify `"./black.pex"` in call sites. I like that part a lot - you remove something that can be error-prone.

--

Upon a re-read of this all, I think the real reason passing an `ExecuteProcessRequest` feels off to me is that you're calling `dataclass.replace` on that, which is not transparent to call sites unless they look up the implementation. When I set `argv` in an `ExecuteProcessRequest`, I would expect exactly that `argv` to be used and it's surprising to me for some values to be injected.

Instead, if you had a kwarg called `extra_args`, that would make it abundantly clear to call-sites that these are being appended to the `argv`, rather than being the whole thing.

As John implemented it, you can use `**kwargs` to avoid having to duplicate every single param of `ExecuteProcessRequest` like `is_nailgunnable`. Afaict, the two things to make explicit are `extra_env` and `extra_args`.